set_sky improvements, set_sun, set_moon and set_stars

Jordach opened the pull request:
Implements Player only controls for the sun, moon, stars and skybox.

- To allow full mod or game control over the default sky colours, sun, moon and stars.
- By adding Player specific functions that allow for fine control over the sun, moon, stars and sky.
- Removes Irrlicht's texture filtering on the sun, moon, to allow for low resolution textures.
- Has full backwards compatibility with the original set_sky call, and hides the sun, moon and stars when used for textured skyboxes.
- Supports changing the colours of the default skybox, day, dawn, night and indoors, as well as sun / moon tinting at dawn and sunrise.

## To do:

Ready for Review.

## How to test

@sofar 's skybox mod: https://content.minetest.net/packages/sofar/skybox/
An experimental version of atmos containing all four calls: [atmos.zip](https://github.com/minetest/minetest/files/3633436/atmos.zip)

## Fancy Screenshots:

(EDIT by paramat: Note these are all of a skybox using 6 supplied textures.)

![screenshot_20190822_014327](https://user-images.githubusercontent.com/1439852/65285728-3cb8bb80-db35-11e9-8cbc-a2608599e839.png)
![screenshot](https://cdn.discordapp.com/attachments/369137254641303560/614085757233266708/unknown.png)
![screenshot](https://cdn.discordapp.com/attachments/369137254641303560/613898060128452631/unknown.png)
![screenshot](https://cdn.discordapp.com/attachments/369137254641303560/613897672021114892/unknown.png)
![screenshot](https://cdn.discordapp.com/attachments/369137254641303560/613897269057552433/screenshot_20190821_174813.png)

lhofhansl commented:
Cool.
Can this blend sky box textures with the day/night colors? I love the textured skybox, but you lose the day/night changes.

Jordach commented:
> Cool.
> Can this blend sky box textures with the day/night colors? I love the textured skybox, but you lose the day/night changes.

Absolutely *not*. You can however, change the ColorSpec of the regular skybox and change those to your will, including the sun and moon tinting at dawn/sunset.

tuedel commented:
It would be great to have an option to disable horizon/"far cloudy fog thing" drawing as well, so it doesn't cover all bodies below the horizon.

The most obvious use case would be space-themed skyboxes.

paramat commented:
If 'set sky' isn't used is the sky completely unchanged in appearence? I seem to remember that was the new approach you chose after closing the previous PR.

Jordach commented:
> Are all default aspects of sky completely unchanged in appearence and behaviour?
It also applies for older mods that use the original set_sky call, not just Minetest's default sky method.

paramat commented:
> It also applies

Sorry, that's unclear, what also applies? And what do you mean? Do you mean that the behaviour and appearence of existing uses of 'set sky' is unchanged?

Jordach commented:
> Sorry, that's unclear, what also applies? And what do you mean? Do you mean that the behaviour and appearence of existing uses of 'set sky' is unchanged?

Running sofar's skybox mod demonstrates this: you'll notice the old call that adds a textured skybox without the sun, moon and stars showing. Which is what the engine does currently, and also applies to the "plain" skybox as well.

Running the embedded atmos experiment shows that the new call while still using a textured skybox, will not disable the sun, moon and stars, unlike the backwards compatible call. Using the older call will also log deprecated.

paramat commented:
Ok thanks.
Please could you answer my questions?:

Are all default aspects of sky, both the default sky and existing usages of 'set sky', completely unchanged in appearence and behaviour?

Are all the issues we discussed in the previous 2 PRs now resolved?

I expect the answers are yes. Sorry to be careful but the last 2 PRs has many issues that went unresolved and unanswered.

paramat commented:
I have a request: If i am discussing something with you in a thread, as above, please post your responses in the thread, not in IRC. Otherwise, the record of the discussion makes no sense in either place.
Also, the discussion is far more accessible here than in IRC logs which few people look at.

My question above wasn't well worded sorry, trying again ...

Is the appearence of sky completely unchanged in both of these situations?:
* The default unmodified sky.
* Existing uses of 'set_sky' in mods.

I understand that the following will cause a slight change, but it is a welcome improvement and i'm not concerned about this:
"Removes Irrlicht's texture filtering on the sun, moon, to allow for low resolution textures."

Jordach commented:
> Is the appearence of sky completely unchanged in both of these situations?:

> * The default unmodified sky.
Yes.
> * Existing uses of 'set_sky' in mods.
Also yes. I ensured the sun, moon and stars hide when using the existing mods to not break visual consistency.

Both situations are covered by the PR.

GreenXenith commented:
Is it possible to get this in before the feature freeze? We (read: a lot of people) wanted this back in 5.0.0 but it didn't happen, and I hope we don't have to wait til 5.2.0 for it.

sfan5 commented on src/client/sky.cpp:
```diff
@@ -205,143 +210,103 @@ void Sky::render()
+						vertex.Pos.rotateXZBy(90);
+					} else if (j == 7) // Left texture
+						vertex.Pos.rotateXZBy(90);
+					else if (j == 8) // Right texture
```
inconsistent braces, all `else`'s should have them

sfan5 commented on src/client/sky.cpp:
```diff
@@ -640,113 +620,218 @@ void Sky::draw_sun(video::IVideoDriver *driver, float sunsize, const video::SCol
+			// appears over the horizon.
+			// Also tune so that stars are at full brightness from time 20000
+			// to time 4000.
+			float starbrightness = MYMAX(0,	MYMIN(1, (0.25 - fabs(wicked_time_of_day < 0.5 ?
```
this could use some reformatting
```cpp
float tod = wicked_time_of_day < 0.5f ? wicked_time_of_day : (1.0f - wicked_time_of_day);
float starbrightness = clamp((0.25f - fabs(tod)) * 20.0f, 0.0f, 1.0f);
```

sfan5 commented on src/client/sky.cpp:
```diff
@@ -640,113 +620,218 @@ void Sky::draw_sun(video::IVideoDriver *driver, float sunsize, const video::SCol
+				vertices[i * 3 + 1].Color = starcolor;
+				vertices[i * 3 + 2].Pos = p2;
+				
+
```
empty lines

sfan5 commented on src/client/sky.cpp:
```diff
@@ -640,113 +620,218 @@ void Sky::draw_sun(video::IVideoDriver *driver, float sunsize, const video::SCol
+		vertices[1] = video::S3DVertex(pos_2, pos_1, -1, 0, 0, 1, c, o, t);
+		vertices[2] = video::S3DVertex(pos_2, pos_2, -1, 0, 0, 1, c, o, o);
+		vertices[3] = video::S3DVertex(pos_1, pos_2, -1, 0, 0, 1, c, t, o);
+	}
```
indented for no reason

paramat replied:
Not resolved, but not an essential change.

Jordach replied:
Should it stay indented or revert to as it was?

sfan5 replied:
Yes, please do fix obvious code style mistakes.

sfan5 commented on src/client/sky.cpp:
```diff
@@ -640,113 +620,218 @@ void Sky::draw_sun(video::IVideoDriver *driver, float sunsize, const video::SCol
+			// Body is directed to -Z (south) by default
+			vertex.Pos.rotateXZBy(horizon_position);
+			vertex.Pos.rotateXYBy(day_position);
+		}
```
same here

paramat replied:
I cannot see any problem here.

sfan5 replied:
The whole function is indented with one more tab than necessary.

sfan5 commented on src/client/sky.h:
```diff
@@ -63,6 +60,127 @@ class Sky : public scene::ISceneNode
+		} else {
+			m_sun_texture = nullptr;
+		}
+	}
```
function too big, move it to sky.cpp (also applies to the all the other ones)

sfan5 commented on src/client/sky.h:
```diff
@@ -63,6 +60,127 @@ class Sky : public scene::ISceneNode
+				// Disables texture filtering
+				m_materials[3].setFlag(video::E_MATERIAL_FLAG(0x100), false);
+				m_materials[3].setFlag(video::E_MATERIAL_FLAG(0x200), false);
+				m_materials[3].setFlag(video::E_MATERIAL_FLAG(0x400), false);
```
these have names

sfan5 commented on src/client/sky.h:
```diff
@@ -63,6 +60,127 @@ class Sky : public scene::ISceneNode
+		} else {
+			//m_sunrise_texture = tsrc->getTextureForMesh("sunrisebg.png");
+			m_materials[2].setTexture(0, tsrc->getTextureForMesh("sunrisebg.png"));
+		}
```
```cpp
m_materials[2].setTexture(0, tsrc->getTextureForMesh(sunglow_texture.empty() ? "sunrisebg.png" : sunglow_texture));
```

sfan5 commented on src/client/sky.h:
```diff
@@ -63,6 +60,127 @@ class Sky : public scene::ISceneNode
+				// Disables texture filtering
+				m_materials[4].setFlag(video::E_MATERIAL_FLAG(0x100), false);
+				m_materials[4].setFlag(video::E_MATERIAL_FLAG(0x200), false);
+				m_materials[4].setFlag(video::E_MATERIAL_FLAG(0x400), false);
```
same

sfan5 commented on src/client/sky.h:
```diff
@@ -79,12 +197,50 @@ class Sky : public scene::ISceneNode
 		m_skycolor = skycolor;
 	}
-	void setBodiesVisible(bool visible) { m_bodies_visible = visible; }
+	void updateSkyColors(video::SColor bg_day, video::SColor bg_dawn,
```
why is this called `update` unlike the others who are called `set`?

Jordach replied:
Perhaps I had a typo, fixed that for consistency.

sfan5 commented on src/client/sky.h:
```diff
@@ -134,25 +290,80 @@ class Sky : public scene::ISceneNode
+	u32 m_star_count = 0;
+
 	video::ITexture *m_sun_texture;
+	video::ITexture *m_sunrise_texture;
```
unused?

Jordach replied:
Confirmed unused, was originally used for something else, but no longer required.

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1698,43 +1698,168 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
+		However, we want to ensure that the skybox can be set to nil when
+		using "regular" or "plain" skybox modes as textures aren't needed.
+		*/
+		if (skybox_params.params.size() != 6 && skybox_params.params[0] != "")
```
then why not just allow size to be either 0 or 6

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1744,26 +1869,250 @@ int ObjectRef::l_get_sky(lua_State *L)
+	// Add textures to a numerically indexed table for clean dump() prints
+	// We also don't want to show it when we're not using the textured skybox
+	if (skybox_params.params[0] == "" || skybox_params.type != "skybox") {
+		lua_pushnil(L);
```
this nil is left over since setfield isn't called

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1744,26 +1869,250 @@ int ObjectRef::l_get_sky(lua_State *L)
+	lua_setfield(L, -2, "moon_tint");
+	lua_pushstring(L, skybox_params.tint_type.c_str());
+	lua_setfield(L, -2, "tint_type");
+	lua_setfield(L, -2, "sky_color");
```
missing value?

Jordach replied:
https://gist.github.com/Jordach/f34d75ee2a75226087e950cdfc76ac49

Dumped output of get_sun/moon/stars/sky.

sfan5 replied:
oh, it's a table

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1744,26 +1869,250 @@ int ObjectRef::l_get_sky(lua_State *L)
+	RemotePlayer *player = getplayer(ref);
+	if (!player)
+		return 0;
+	if (lua_istable(L, 2)) {
```
should be early-exit instead

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1744,26 +1869,250 @@ int ObjectRef::l_get_sky(lua_State *L)
+	RemotePlayer *player = getplayer(ref);
+	if (!player)
+		return 0;
+	if (lua_istable(L, 2)) {
```
same

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1744,26 +1869,250 @@ int ObjectRef::l_get_sky(lua_State *L)
+		star_params.count = getintfield_default(L, 2,
+			"count", star_params.count);
+
+		star_params.starcolor = video::SColor(255, 235, 235, 255);
```
why do default values (such as this one) sometimes appear here?
`RemotePlayer::RemotePlayer()` sets defaults, why does keeping them not work?

Jordach replied:
Defaults are returned from already set values - ie, mod supplied values. If any part was set to `nil`: it'd keep the previously last used value.

See: set_clouds.

sfan5 replied:
This should not be necessary, `starcolor` will always be set to default if left alone.
If stuff segfaults if you don't do this, the problem is elsewhere and this code is just bandaid.

Jordach replied:
I believe this was done in the original set_sky, to prevent invalid "colours" being set.

Jordach replied:
`read_color()` has a built in protection which prevents invalid colours like #1234567890

paramat replied:
@sfan5 is this resolved in your opinion?

paramat replied:
@Jordach please do not resolve discussions like this, they can only be resolved if the core dev is satisfied.

sfan5 replied:
Not resolved. I'd like a concrete case where removing this line breaks stuff.

sfan5 commented on src/server.cpp:
```diff
@@ -1671,17 +1671,56 @@ void Server::SendHUDSetParam(session_t peer_id, u16 param, const std::string &va
-	for (const std::string &param : params)
-		pkt << param;
+	pkt << params.bgcolor << params.type
+		<< params.clouds;
```
there's zero backwards-compatibility here

Jordach replied:
What if the server sent a TOCLIENT_SET_SUN or TOCLIENT_SET_STARS packet? That would cause a client error trying to figure out said packet. API compatible was a better choice than packet compatible.

Jordach replied:
Now backwards compatible upto 5.0.[0/1] clients.

sfan5 commented on src/skyparams.h:
```diff
@@ -0,0 +1,64 @@
+    video::SColor bgcolor;
+    std::string type;
+    std::vector<std::string> params;
+    bool clouds;
```
wrong indent (also in the other structs)

Jordach replied:
`cloudparams.h` does the exact same thing.

sfan5 replied:
Then it's also wrong.

paramat replied:
Indentation needs fixing.

paramat replied:
Resolved.

sfan5 commented on src/network/clientpackethandler.cpp:
```diff
@@ -1233,28 +1233,129 @@ void Client::handleCommand_HudSetParam(NetworkPacket* pkt)
+			>> dawn_sky >> dawn_horizon
+			>> night_sky >> night_horizon
+			>> indoors;
+	}
```
for easier forwards-compatibility the client should abort (or do something other useful) when encountering an unknown type here

LoneWolfHT commented:
Agreed, it would be really nice if we could get this in before the feature freeze.
I unfortunately can't test until around the time Ludum Dare starts up their voting

paramat commented:
Jordach, thanks.
> Is it possible to get this in before the feature freeze?

> Agreed, it would be really nice if we could get this in before the feature freeze.

That would probably be unacceptably rushing it, it's a big feature PR that is likely to need consideration and improvement (see the review above), and we're busy attending to the priorities for releasing 5.1.0.
It's a very welcome feature and i will certainly carefully review and test this after 5.1.0 release.
> We (read: a lot of people) wanted this back in 5.0.0 but it didn't happen

Keep in mind the previous 2 PRs never reached a mergable state and were abandoned, so it's not as if there was something ready to merge, and the delay wasn't due to core dev neglect.
Also of course, there are many things a lot of people have wanted for a long time, this is no different, there is no justification for rushing a merge, or rushing a merge for this in particular. But i think you know all this =)

Jordach commented on doc/lua_api.txt:
```diff
@@ -5668,15 +5668,87 @@ This is basically a reference to a C++ `ServerActiveObject`
+* `set_sky(parameters)`
+    * `parameters` is a table with the following optional fields:
+        * `base_color`: ColorSpec, changes fog in certain skybox modes.
+        * `type`: Available types:
```
Like the prior set_sky, `base_color` is the same as `bgcolor` in the older, deprecated call, and are treated identically.

Jordach commented on src/client/sky.cpp:
```diff
@@ -67,44 +67,53 @@ Sky::Sky(s32 id, ITextureSource *tsrc):
 		m_materials[4].setTexture(0, m_moon_texture);
 		m_materials[4].MaterialType = video::EMT_TRANSPARENT_ALPHA_CHANNEL;
+		// Disables texture filtering
+		m_materials[4].UseMipMaps = false;
```
Wasn't needed after all: mipmaps only apply to far-distant objects to "blur them". The skybox is never considered far from the camera, despite being rendered "far" visually.

Jordach commented on doc/lua_api.txt:
```diff
@@ -5668,15 +5668,87 @@ This is basically a reference to a C++ `ServerActiveObject`
+            will re-enable the mesh sun. (default: `"sun.png"`)
+        * `tonemap`: A 512x1 texture containing the tonemap for the sun
+            (default: `"sun_tonemap.png"`)
+        * `sunrise`: String 
```
* `sunrise`: A regular texture for the sunrise texture. (default: `"sunrisebg.png"`)

Was supposed to contain this, now included.

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1698,43 +1698,167 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
+		lua_getfield(L, 2, "textures");
+		skybox_params.params.clear();
+		if (!lua_istable(L, -1)) {
+			//skybox_params.params.emplace_back("");
```
should be removed fully

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1744,26 +1868,249 @@ int ObjectRef::l_get_sky(lua_State *L)
+	// Add textures to a numerically indexed table for clean dump() prints
+	// We also don't want to show it when we're not using the textured skybox
+	if (skybox_params.params.size() != 6 || skybox_params.type != "skybox") {
+	} else {
```
should be
```cpp
if (skybox_params.params.size() == 6 && skybox_params.type == "skybox") {
```

paramat replied:
Not resolved.
I agree with sfan5.

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1744,26 +1868,249 @@ int ObjectRef::l_get_sky(lua_State *L)
+		"scale", moon_params.scale);
+		
+	getServer(L)->setMoon(player, moon_params);
+	return 1;
```
missing `lua_pushboolean(L, true);`

paramat replied:
Was this resolved?

sfan5 replied:
nope, bug still present

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1744,26 +1868,249 @@ int ObjectRef::l_get_sky(lua_State *L)
+		"size", star_params.scale);
+
+	getServer(L)->setStars(player, star_params);
+	return 1;
```
missing `lua_pushboolean(L, true);`

paramat replied:
Was this resolved?

sfan5 replied:
same here, not resolved

sfan5 commented on src/server.cpp:
```diff
@@ -1671,17 +1671,68 @@ void Server::SendHUDSetParam(session_t peer_id, u16 param, const std::string &va
-	for (const std::string &param : params)
-		pkt << param;
+	// Handle prior clients here
+	if (m_clients.getProtocolVersion(peer_id) < 38) {
```
this should be `< 39` and the version should be bumped in networkprotocol.h

Jordach replied:
Network protocol bumped to 39; should min version be bumped as 5.0.0 clients connect fine?

ClobberXD replied:
AFAIK, min should be left as-is to prevent breakage of backwards compat with 5.0.x clients (hopefully this PR doesn't break existing the `set_sky` in any way). Bumping the min version would mean that this PR would remain unmerged until 6.0.0 :P

sfan5 replied:
No, only `LATEST_PROTOCOL_VERSION` needs to be bumped.

Jordach replied:
@ClobberXD it's API and packet compatible if older versions (of the Lua call, and client version)  are detected.

sfan5 replied:
Not resolved.

sfan5 commented on src/server.cpp:
```diff
@@ -1671,17 +1671,68 @@ void Server::SendHUDSetParam(session_t peer_id, u16 param, const std::string &va
+	if (m_clients.getProtocolVersion(peer_id) < 38) {
+		pkt << params.bgcolor << params.type << (u16) params.params.size();
+
+		for(size_t i=0; i<params.params.size(); i++)
```
`for (const std::string &texture : params.params)`

sfan5 replied:
Also not resolved.

sfan5 commented on src/server.cpp:
```diff
@@ -1671,17 +1671,68 @@ void Server::SendHUDSetParam(session_t peer_id, u16 param, const std::string &va
+
+		pkt << params.clouds;
+
+		Send(&pkt);	
```
`Send()` is called outside the if, this one is duplicate

Jordach replied:
Ah. I missed that one.

paramat replied:
Seems unresolved?

Jordach replied:
Mistake was resolved when I added backwards compatibility.

paramat replied:
Now fixed.

paramat commented:
I've worked on sky code a lot, please don't merge this until i've reviewed and tested. There isn't time to do that before freeze, but i'll make this my highest priority for review (below the priority of any work needed for releasing 5.1.0).

paramat commented:
![screenshot](https://cdn.discordapp.com/attachments/369137254641303560/613897269057552433/screenshot_20190821_174813.png)

Horizon

Is this a screenshot of the PR? This illustrates a problem the last PR had.
The horizon present in the standard MT sky, that obscures sun moon and stars, is missing and stars are visible far below the horizon just beyond the loaded mapblocks, this looks very bad.
Imagine the player is on a mountain looking down at a limited area of loaded world around the base of the mountain, it seems stars will be visible looking downwards, so this problem will be even worse.
Is it just stars that are not obscured below the horizon, or are sun and moon also not obscured?

The standard obscuring horizon creates the impression of more distant world in the absence of loaded mapblocks.
It also creates a 'horizon line' that acts as the visual orientation line to give a sense of horizontality and view pitch. Without it the world becomes disorientating, the horizon moves around dependent on loaded mapblocks. In the screenshot, the terrain and building looks tilted towards the player due to the incorrect horizon.
It would be good for an obscuring horizon to be switchable for space mods/games. Please can an optional obscuring horizon be added? As discussed before it should be placed at 0 degrees view pitch.


Far cloudy fog thing

The parameters suggest that colours are settable for upper and lower sky to create a colour gradient, i have no problem with that it's a good feature. But where are te parameters for the standard colour band just above the horizon, the 'far cloudy fog thing', which extends downwards to cover the lower half of the skybox?
If i want to customise sky i want that to be an option and want to be able to customise its colour. Please can that be added? It would be good for that to be switchable for space mods/games.

////////////////////////////////

Both the above issues were issues that were present in the last PR, which is why i asked whether the issues of the last PRs have been resolved.

An API to customise sky means the ability to customise the current standard appearence of sky, with perhaps some welcome additional options. But a customised standard appearence is not possible here.
This is an optional new sky implementation, which might be acceptable, but then it shouldn't be presented as a customisable standard sky.
If merged as it is, additional APIs will have to be created to allow a customised standard sky with horizon and 'far cloudy fog thing', this could get messy.

I know you don't like the standard sky, but a PR to customise sky has to allow a customised standard sky identical in appearence. If you don't want to code a customisable standard sky then the API should be designed to make clear this is an optional 'new sky' implementation, in which case it can't have the form:
'set sky' -> parameters
... because that implies a customisable standard sky, and it would get in the way of APIs for a customisable standard sky.
A new sky implementation doesn't necessarily need a 'far cloudy fog thing' because you have a gradient instead, however, it does require the possibility of an obscuring horizon.

:-1: for current state of the PR but with significant changes will of course be very welcome.

LoneWolfHT commented:
I'm personally fine with the standard sky changing. From what I've seen it will be changing for the better.
Visible changes are nice too. It gives a feeling of progress to the younger players who don't care about or follow the code

paramat commented:
> I'm personally fine with the standard sky changing

Ok, that is fine, but that ignores the problems i explained.
Without an obscuring horizon it isn't a change for the better.
We could have an optional new sky appearence, that would be fine but this PR doesn't do that in an acceptable way.

Jordach commented:
@paramat those screenshots are taken with a *textured skybox*

![screenshot_20190925_202656](https://user-images.githubusercontent.com/1439852/65632949-d7504900-dfd2-11e9-813d-1b1238d4113e.png)

paramat commented:
To clarify:
I think the ideal form of this PR is a customisable classic sky plus options that create the new sky appearence, because then we get both within 1 API.
This makes sense because the only additional feature of Jordach's preferred sky appearence is the colour gradient.
Users can set both colours of the gradient identical to create a classic sky with custom colours.

So my requests are:
* Re-add (unchanged) the 'far cloudy fog thing' and the obscuring horizon as a option, default enabled, and switch both with an added bool in parameters.
Disabling them allows a sky suitable for space games. Making these optional has actually been requested and considered before, so will be another feature request closed.
* Add a colour parameter for the 'far cloudy fog thing'

With this i will continue to review and, if no other issues arise, will probably approve.

sfan5 commented:
> Re-add (unchanged) the 'far cloudy fog thing' and the obscuring horizon

Have you actually tested this PR? Both of these features are still present.

Jordach commented:
I made sure that still worked, will look at fixing this evening.

> On 26 Sep 2019, at 17:04, sfan5 <notifications@github.com> wrote:
> 
> ﻿
> Just noticed something broken while testing: The inside-clouds fog that was added with #6003 no longer works.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

paramat commented:
> Have you actually tested this PR? Both of these features are still present.

Not yet. They may be present for the default sky, i understand that, but it seems they are not possible for the custom sky, that is what i mean. Please see my long review post for clarification.

The screenshots have no horizon or 'far cloudy fog thing' and the documented parameters do not have anything that would either switch those or define the colour of the 'far cloudy fog thing'. So it seems clear those have been removed for the custom sky.

paramat commented:
> those screenshots are taken with a textured skybox

As in 6 provided textures? Ok i may have misunderstood. However you should have provided screenshots for all uses, many people will assume that what is in the screenshots is all that is possible.
Also, where is the colour control for the 'far cloudy fog thing'? Surely that is necessary?

paramat commented:
Useful templates for testing (from the atmos zip file):
```
	player:set_sky({
			base_color = "#ff0000",--minetest.rgba(
				--result_fog.r,
				--result_fog.g,
				--result_fog.b),
			type = "regular",
			textures = {
				"sky_tex_top",
				"sky_tex_bottom",
				"sky_tex_side1",
				"sky_tex_side2",
				"sky_tex_side3",
				"sky_tex_side4"
			},
			clouds = true,
			sky_color = {
				day_sky = "#ff00ff",
				day_horizon = "#88ff00",
				dawn_sky = "#444400",
				dawn_horizon = "#ffff00",
				night_sky = "#000044",
				night_horizon = "#000088",
				indoors = "#888888",
				sun_tint = "#00ffff",
				moon_tint = "#ff0000",
				tint_type = "custom"
			}
		})

	player:set_sun({
			visible = true,
			texture = "atmos_sun.png",
			--tonemap = nil,--"test",
			--sunrise = "sunrisebg.png",--"tset",
			--sunrise_visible = true,
			rotation = 90,
			scale = 2
		}
	)
	
	player:set_moon({
			visible = true,
			texture = "atmos_moon.png",
			tonemap = "",
			rotation = -90,
			scale = 5
		}
	)
	
	player:set_stars({
		visible = true,
		count = 2500,
		star_color = minetest.rgba(235, 235, 255, 140),
		rotation = 0,
		size = 0.5,
	})
```

paramat commented on doc/lua_api.txt:
```diff
@@ -5668,15 +5668,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+            (default: `"sun_tonemap.png"`)
+        * `sunrise`: A regular texture for the sunrise texture.
+            (default: `"sunrisebg.png"`)
+        * `sunrise_visible`: Boolean for whether the sunrise is visible.
```
Boolean for whether the sunrise texture is visible.

paramat commented:
```
        * `rotation`: Floating point value to set what direction the
            stars should rotate around the world. (default: `0`)
```
This does not seem necessary as stars are randomly placed so altering the rotation phase will not have any noticeable effect.

paramat commented on doc/lua_api.txt:
```diff
@@ -5668,15 +5668,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+        * `type`: Available types:
+            * `"regular"`: Uses 0 textures, `base_color` ignored
+            * `"skybox"`: Uses 6 textures, `base_color` used as fog.
+            * `"plain"`: Uses 0 textures, `base_color` used as fog.
```
"base_color used for plain sky color and fog."?

paramat commented on doc/lua_api.txt:
```diff
@@ -5668,15 +5668,88 @@ This is basically a reference to a C++ `ServerActiveObject`
-* `get_sky()`: returns bgcolor, type, table of textures, clouds
+* `set_sky(parameters)`
+    * `parameters` is a table with the following optional fields:
+        * `base_color`: ColorSpec, changes fog in certain skybox modes.
```
Also changes the sky color in "plain" mode.

paramat commented:
I can't seem to get the "plain" skybox working, is my code wrong?:
```
		player:set_sky({
			base_color = "#ff0000",
			type = "plain",
			clouds = true,
		})
```

paramat commented:
Testing with:
```
		player:set_sky({
			type = "regular",
			textures = {},
			clouds = true,
			sky_color = {
				day_sky = "#ff00ff", -- magenta
				day_horizon = "#00ff00", -- green
				dawn_sky = "#ffff00", -- yellow
				dawn_horizon = "#00ffff", -- cyan
				night_sky = "#ff0000", -- red
				night_horizon = "#0000ff", -- blue
				indoors = "#888888", -- grey
			}
		})
```
Shows that 'sky' and 'horizon' colours have been flipped for all 3 times:

![screenshot_20191013_213329](https://user-images.githubusercontent.com/3686677/66722113-9db87400-ee01-11e9-8fe7-ccb7a73f3716.png)

![screenshot_20191013_213211](https://user-images.githubusercontent.com/3686677/66722118-a1e49180-ee01-11e9-9d05-5489c4a1ca3c.png)

![screenshot_20191013_213307](https://user-images.githubusercontent.com/3686677/66722119-a6a94580-ee01-11e9-9de6-8db1a8444069.png)

paramat commented on doc/lua_api.txt:
```diff
@@ -5668,15 +5668,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+            (default: `true`)
+        * `rotation`: Floating point value to set what direction the
+            sun should rotate around the world. (default: `90`)
+        * `scale`: Float controlling the overall size of the sun. (default: `1`)
```
Sun texture only?

paramat commented on doc/lua_api.txt:
```diff
@@ -5668,15 +5668,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+            (default: `"moon_tonemap.png"`)
+        * `rotation`: Floating point value to set what direction the
+            moon should rotate around the world. (default: `-90`)
+        * `scale`: Float controlling the overall size of the moon (default: `1`)
```
For a moon texture only?

paramat commented:
There is strange behaviour with moon 'rotation', if '80' is used the moon appears in daytime to one side of the sun and rises/sets with the sun. If '-45' is used it appears at nighttime as expected.
I assume the intention was that sun and moon rise/set times are not changed and that rotation just offsets the sun/moon Northwards aor Southwards?

paramat commented:
![screenshot_20191013_221558](https://user-images.githubusercontent.com/3686677/66722571-461d0700-ee07-11e9-9d6a-d75a128086a7.png)

"skybox" mode using textures and red fog colour,at midnight.
Works but the sunrise texture is now always visible just below horizon level because there is no obscuring horizon to hide it. So i think it should be force-disabled in "skybox" mode.

paramat commented:
I'm haivng issues with the old 'set_sky' call.
```
		player:set_sky(
			"#000000",
			"skybox",
			{
				"default_sand.png",
				"default_stone.png",
				"default_grass.png",
				"default_jungletree.png",
				"default_obsidian.png",
				"default_lava.png"
			},
			false
		)
```
The first time it is called it works, the 2nd time it is called it causes a crash with:
`ERROR[Main]: ServerError: AsyncErr: environment_Step: Skybox expects 6 textures.`

Also, the old call does not work when using a colorspec in the form `{r = 0, g = 0, b = 0, a = 0}`.

paramat commented:
Apart from all that, everything i tested works.
I have not yet tested anything related to 'directional colored fog', or custom sun/moon/sunrise textures or tonemaps.

paramat commented:
@Jordach note that i have removed my disapproval which was due to my misunderstanding. With a few things fixed and possibly some minor changes this will probably be mergeable, it is an excellent feature.

Jordach commented:
Thanks, my PC has been inoperable due to unexpected hardware failure (and thus unable to make changes), I believe the old call has an issue when pulling ColorSpec with RGB+Alpha, which is why it defaulted to pure white as a fallback colour (and testing reveals that it’s unnecessary as it will use the last entered value, or black if not given prior data).

If you have any testing code, for the prior set_sky call, I’d be glad to fix it properly.

> On 18 Oct 2019, at 21:44, Paramat <notifications@github.com> wrote:
> 
> @Jordach <https://github.com/Jordach> note that i have removed my disapproval which was due to my misunderstanding. With a few things fixed and possibly some minor changes this will probably be mergeable, it is an excellent feature.
>

Jordach commented:
The peculiarity with the old call appears to be a (possible regression) with `read_color()` not getting the table properly. Will test locally.

paramat commented:
> Re: Cloud inside visibility; turns out the cloud function doesn't appear to take in cloud thickness into account.

See #8340 for more information on this, it turns out to be due to camera offset.

paramat commented:
>  I had considered adding an offset, in that the sunrise's and moonrise's initial start positions can be fiddled with. Along with extra pitch/roll. My current thoughts at 3AM indicate adding a table akin to set_clouds() speed table, allowing for
set_sun/moon/stars.transforms.init_pitch/roll/yaw.

Rotation is redundant for stars as they are uniformly distributed.
3 rotations for sun and moon is unnecessary, only 2 are needed to place them anywhere in the sky.
Also, rotating the sun along the path of its movement (altering rise/set times) makes no sense because day/night and lighting is all dependent on it staying where it is. It can only be offset East or West as currently (if fixed).

So at most we could have:
Stars: No rotations.
Sun: 1 rotation, offset to East or West, as currently (but fixed).
Moon: 2 rotations: 1 along path of movement  (altering rise/set times). 1 offset to East or West, as currently (but fixed).

paramat commented:
@Jordach still having PC problems?
With some changes and fixes this will be very welcome.

pyrollo commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1698,43 +1698,167 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
+			lua_pop(L, 1);
+			getServer(L)->setSky(player, skybox_params);
+			lua_pushboolean(L, true);
+		}
```
Seems that ```getServer(L)->setSky``` and ```lua_pushboolean``` calls are misplaced. They should be outside this block so they are also called when no "```sky_color```" table given. I had to change that to make following Lua code work:
```lua
player:set_sky({base_color = "#f1f4ee", type = "skybox", clouds = false, textures = {"Up.jpg", "Down.jpg", "Front.jpg", "Back.jpg", "Left.jpg", "Right.jpg"} } )
```

Jordach commented:
Rebased against 5.2.0: no issues.

Fixed @pyrollo 's issues:

Fixed ColorSpec Tables and the regular table colliding due to is_table():

```
lua_getfield(L, 2, "r");
is_colorspec = lua_isnil(L, -1) ? false : true;
```

rubenwardy commented:
What needs to be done on this now?

paramat commented:
I will retest for the bugs i found earlier.

Jordach commented:
@paramat status on bug testing?

paramat commented:
I have been too busy, but rereviewing this is high priority for me.

paramat commented:
In the new commits i cannot see a fix of https://github.com/minetest/minetest/pull/8964#issuecomment-541457239 or did i miss it?

paramat commented:
First bug in https://github.com/minetest/minetest/pull/8964#issuecomment-541462472 still occurs.

paramat commented:
2nd bug of https://github.com/minetest/minetest/pull/8964#issuecomment-541462472 has now changed to this:

Using the old set_sky call with code:
```
player:set_sky(
			{r = 0, g = 0, b = 0, a = 0},
			"skybox",
			{
				"default_sand.png",
				"default_stone.png",
				"default_grass.png",
				"default_jungletree.png",
				"default_obsidian.png",
				"default_lava.png"
			},
			false
		)
```
It works the first time, but crashes the second time with error:
`ERROR[Main]: ServerError: AsyncErr: environment_Step: Skybox expects 6 textures.`
So the behaviour and error are now identical to the first bug of https://github.com/minetest/minetest/pull/8964#issuecomment-541462472

Jordach commented:
I found the bug in question with the expects six textures:

The storage of the textures was not cleared on another old set_sky call, leading to 12 textures being kept in memory, rather than being cleared.

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -5832,15 +5832,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+            * `"plain"`: Uses 0 textures, `base_color` used as both fog and sky.
+        * `textures`: A table containing up to six textures in the following
+            order: top, bottom, left, right, front, back.
+        * `clouds`: Boolean for whether clouds appear in front of 
```
Surely this applies also when `type != skybox`, or that a current limitation?

Jordach replied:
`"regular"` gets fog from the current time of day and defined sky colors.

`"plain"` uses it for both the sky and singular sky color.

`"skybox"` only uses it for fog.

rubenwardy replied:
this is a comment on the `clouds` option

Jordach replied:
It applies to all three skybox modes.

paramat replied:
I see, 'in front of the skybox' implies it only works in skybox mode, this would be better as 'in front of the sky'.

rubenwardy commented on src/client/game.cpp:
```diff
@@ -2744,41 +2750,111 @@ void Game::handleClientEvent_HudChange(ClientEvent *event, CameraOrientation *ca
+			*event->set_sky.tint_type
+		);
+		// Clean up related data:
+		delete event->set_sky.day_sky;
```
this is really bad, manual memory management should be avoided

rubenwardy commented on src/client/sky.h:
```diff
@@ -134,25 +155,76 @@ class Sky : public scene::ISceneNode
-	v3f m_stars[SKY_STAR_COUNT];
+	
+	// Horizon colors
+	video::SColorf m_bgcolor_day_f = video::SColor(255, 155, 193, 240);
```
I highly recommend separating state from drawing logic. By this, I mean that this should have a separate class called SkyParams which contains all the stuff needed by set_sky, and an object called SkyObjectParams which contains the stuff needed by set_moon and set_sun (one contains the subset of another). This would simplify the gross code in client/event.

rubenwardy commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1701,43 +1701,175 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
+		lua_pop(L, 1);
+
+		lua_getfield(L, 2, "textures");
+		skybox_params.params.clear();
```
`params` should be `textures`

rubenwardy commented on src/client/clientevent.h:
```diff
@@ -153,6 +156,16 @@ struct ClientEvent
 			std::string *type;
 			std::vector<std::string> *params;
 			bool clouds;
+			video::SColor *day_sky;
```
gross - pointers here are largely unnecessary. If you want to reduce the size of the uninon, use `std::unique_ptr`

rubenwardy commented:
please don't resolve threads until you've fixed them, it makes it hard to keep track of things

paramat commented:
Jordach, great, all my issues are probably resolved now then, will retest anyway.

In case it is useful my testing code for 2 of those bugs is (mod should depend on default):
```
local function cyclic_update()
	for _, player in ipairs(minetest.get_connected_players()) do
		player:set_sky(
			"#000000",
			--{r = 0, g = 0, b = 0, a = 0},
			"skybox",
			{
				"default_sand.png",
				"default_stone.png",
				"default_grass.png",
				"default_jungletree.png",
				"default_obsidian.png",
				"default_lava.png"
			},
			false
		)
		print("set sky")
	end
	minetest.after(4.7, cyclic_update)
end

minetest.after(4.7, cyclic_update)
```

Jordach commented:
There, that should get everything, @paramat I'm ready for a second review to double check I didn't miss anything.

paramat commented:
I will retest everything.

paramat commented:
Clang Format reports indentation problems (looks like spaces instead of tabs) in the structs in src/skyparams.h.
And in src/remoteplayer.h it suggests:
```
-	void setMoon(const MoonParams &moon_params) {m_moon_params = moon_params; }
+	void setMoon(const MoonParams &moon_params) { m_moon_params = moon_params; }
```
However, the CF suggestion:
```
-	void setSky(const SkyboxParams &skybox_params) { m_skybox_params = skybox_params; }
+	void setSky(const SkyboxParams &skybox_params)
+	{
+		m_skybox_params = skybox_params;
+	}
```
seems to be contradictory and a bad suggestion.

Once this PR is approved, any nonsense complaints by CF should be silenced by adding the corresponding files to the whitelist https://github.com/minetest/minetest/blob/master/util/travis/clang-format-whitelist.txt

paramat commented on doc/lua_api.txt:
```diff
@@ -5832,15 +5832,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+    * `parameters` is a table with the following optional fields:
+        * `visible`: Boolean for whether the stars are visible.
+            (default: `true`)
+        * `count`: Interger number to set the number of stars in 
```
'Integer'

paramat replied:
Not resolved. Why did you resolve this? Did you resolve it before making changes then forget?

Jordach replied:
Oops, I was checking them off as I made the changes locally.

paramat replied:
Now fixed.

paramat commented on doc/lua_api.txt:
```diff
@@ -5832,15 +5832,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+              at sunrise and sunset.
+            * `moon_tint`: ColorSpec, changes the fog tinting for the moon
+              at sunrise and sunset.
+            * `tint_type`: ColorSpec, changes which mode the directional fog
```
To be more descriptive, i suggest these field names:
fog_tint_sun
fog_tint_moon
fog_tint_type

paramat replied:
New field names are fine.

paramat commented on doc/lua_api.txt:
```diff
@@ -5832,15 +5832,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+            (default: `"sunrisebg.png"`)
+        * `sunrise_visible`: Boolean for whether the sunrise texture is visible.
+            (default: `true`)
+        * `rotation`: Floating point value to set what direction the
```
Needs removing

paramat commented on doc/lua_api.txt:
```diff
@@ -5832,15 +5832,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+            will re-enable the mesh moon. (default: `"moon.png"`)
+        * `tonemap`: A 512x1 texture containing the tonemap for the moon
+            (default: `"moon_tonemap.png"`)
+        * `rotation`: Floating point value to set what direction the
```
Needs removing

paramat commented on doc/lua_api.txt:
```diff
@@ -5832,15 +5832,88 @@ This is basically a reference to a C++ `ServerActiveObject`
+        * `star_color`: ColorSpec, sets the colors of the stars,
+            alpha channel is used to set overall star brightness.
+            (default: `#ebebff69`)
+        * `rotation`: Floating point value to set what direction the
```
Needs removing

paramat commented on src/client/game.cpp:
```diff
@@ -2744,41 +2750,93 @@ void Game::handleClientEvent_HudChange(ClientEvent *event, CameraOrientation *ca
 	sky->setVisible(false);
-	// Whether clouds are visible in front of a custom skybox
-	sky->setCloudsEnabled(event->set_sky.clouds);
+	// Whether clouds are visible in front of a custom skybox.
```
Comment is incorrect, cloud bool works in any mode

paramat commented on src/client/sky.cpp:
```diff
@@ -67,44 +68,51 @@ Sky::Sky(s32 id, ITextureSource *tsrc):
-			myrand_range(-10000, 10000)
-		);
-		star.normalize();
+	setStarCount(200);
```
See #9307 we decided on a new default of 1000

paramat commented on src/client/sky.cpp:
```diff
@@ -717,14 +702,108 @@ void Sky::draw_moon(video::IVideoDriver *driver, float moonsize, const video::SC
+			float starbrightness = clamp((0.25f - fabs(tod)) * 20.0f, 0.0f, 1.0f);
+			
+			float f = starbrightness;
+			float d = (0.007 / 2) * m_star_params.scale;
```
See #9307 we decided on a new default of (0.006 / 2)

paramat commented on src/remoteplayer.cpp:
```diff
@@ -66,6 +66,33 @@ RemotePlayer::RemotePlayer(const char *name, IItemDefManager *idef):
+	m_moon_params.scale = 1;
+
+	m_star_params.visible = true;
+	m_star_params.count = 200;
```
See #9307 we decided on a new default of 1000

paramat commented on src/remoteplayer.h:
```diff
@@ -83,23 +84,21 @@ class RemotePlayer : public Player
+
+	const SunParams &getSunParams() const { return m_sun_params; }
+
+	void setMoon(const MoonParams &moon_params) {m_moon_params = moon_params; }
```
{ m_moon_params = moon_params; }

paramat commented:
Also, i unresolved a few line comments from a few months back. Some were marked as resolved but are still unresolved or possibly unresolved.

Jordach commented:
> seems to be contradictory and a bad suggestion.

That seems like it's attempting to keep it within style guidelines and under 80 characters.

Jordach commented:
![screenshot_20200115_204954](https://user-images.githubusercontent.com/1439852/72470374-9e81ed00-37d8-11ea-9691-4371b395ac9b.png)
Screenshot of the new stars setting with no set_[sky/sun/moon/stars] used.

paramat commented:
> That seems like it's attempting to keep it within style guidelines and under 80 characters.

Possibly, but up to 90 columns (with tab size 4) is the actual MT guideline. So if it is 90 or less it is fine not wrapped.
I suspect CF does not use our 90 column guideline.

paramat commented:
> Oops, I was checking them off as I made the changes locally.

I have to make a similar post to rubenwardy's ...

Please do not resolve line comments until the change is made exactly as requested and pushed to Github, not before pushing.
From everyone elses point of view, until it is pushed to Github the line comment is not resolved. There can be quite some time between you resolving locally and it being pushed.
If in doubt, for example making a change in a different way to requested, leave a line comment open and a core dev will resolve it.

I discovered several line comments which you had marked as resolved before making the change and then, i assume, forgot to make the change. This wastes review, makes review much more difficult and results in unfixed code errors.
This is very irritating, sorry, if it continues i will refuse to review or test.

paramat commented:
I have unresolved some outdated line comments too, see earlier in the thread for those, they need to be checked by a core dev to see if they are resolved.

Jordach commented:
Would it be worthwhile reverting #9307? Because this PR also contains the same parameters.

paramat commented:
I was surprised that was merged, we were somewhat happy to delay it. However, i do not think reverting it is justified, this should be fairly easy to rebase. Also, this PR is far from ready.

sfan5 commented on src/client/sky.cpp:
```diff
@@ -717,14 +702,108 @@ void Sky::draw_moon(video::IVideoDriver *driver, float moonsize, const video::SC
+				indices, m_star_params.count, video::EVT_STANDARD,
+				scene::EPT_QUADS, video::EIT_16BIT);
+		delete[] indices;
+		delete[] vertices;
```
function body still indented too much

Jordach replied:
Fixed

sfan5 commented on src/client/sky.cpp:
```diff
@@ -737,16 +816,200 @@ void Sky::draw_sky_body(std::array<video::S3DVertex, 4> &vertices, float pos_1,
+	ITextureSource *tsrc) {
+	// Keep a list of texture names handy.
+	m_sky_params.textures.emplace_back(texture);
+	std::cout << texture << std::endl;
```
leftover debug output

Jordach replied:
Fixed

sfan5 commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1701,43 +1701,182 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
+		getServer(L)->setSun(player, sun_params);
+		getServer(L)->setMoon(player, moon_params);
+		getServer(L)->setStars(player, star_params);
+	}
```
missing a `lua_pushboolean(L, true);` here

Jordach replied:
Fixed

sfan5 commented:
Aside from the new review comments there are 5 older ones from me that have not been resolved.
Rebasing (because of #9307) can be done at time of merging IMO.

Jordach commented:
Recent push should fix remaining review notes. Feel free to resolve at your own pace.

sfan5 commented on src/server.cpp:
```diff
@@ -1718,17 +1718,62 @@ void Server::SendHUDSetParam(session_t peer_id, u16 param, const std::string &va
+	if (m_clients.getProtocolVersion(peer_id) < 39) {
+		pkt << params.bgcolor << params.type << (u16) params.textures.size();
+
+		for(size_t i=0; i<params.textures.size(); i++)
```
you missed this one, not essential though

`for (const std::string &texture : params.params)`

kilbith replied:
This is stupid, you can fix it by yourself

paramat replied:
We do not have time to fix contributors' PRs. It is not our responsibility or obligation.
Fixing something ourselves only makes sense if it is certain that it is the very last change needed, and the PR is close to being approved, neither of which is the case here.

This seems resolved, but only sfan5 should resolve this comment.

sfan5 commented:
@paramat any remaining issues from your side?

paramat commented:
sfan5 seems mostly happy with this now so i will retest.

Jordach commented:
The networking for PROTOCOL 38 clients works as expected, as do PROTOCOL 39 clients with a networked demonstration I performed last night.

paramat commented:
Currently testing, updating here as i go.

Custom fog moon / sun tint creates pulatsing fog tint at dusk/dawn.
Try:
```
			sky_color = {
				fog_sun_tint = "#ff0000",
				fog_moon_tint = "#00ff00",
				fog_tint_type = "custom"
			}
```
Make sure 'directional coloured fog' is enabled.
At dusk/dawn, look at both horizons, but it is clearer for moon tint. At certain times, the fog moon tint pulsates from dark to bright green,

paramat commented:
Everything else works fine, including the old set_sky call.

Here's my testing code, a mod depending on 'default':
```local function cyclic_update()
	for _, player in ipairs(minetest.get_connected_players()) do
		print("set sky")
		--[[
		player:set_sky({
			base_color = "#000000",
			--base_color = {r = 0, g = 0, b = 0, a = 0},
			type = "regular",
			--type = "skybox",
			--type = "plain",
			textures = {
				"default_sand.png",
				"default_stone.png",
				"default_grass.png",
				"default_jungletree.png",
				"default_obsidian.png",
				"default_lava.png"
			},
			clouds = true,
			sky_color = {
				--day_sky = "#ff00ff",
				--day_horizon = "#00ff00",
				--dawn_sky = "#00ffff",
				--dawn_horizon = "#ffff00",
				--night_sky = "#0000ff",
				--night_horizon = "#ff0000",
				--indoors = "#ffffff",
				--fog_sun_tint = "#ff0000",
				--fog_moon_tint = "#00ff00",
				--fog_tint_type = "custom"
			}
		})--]]

		--[[
		player:set_sun({
			visible = true,
			--texture = "default_stone.png",
			--tonemap = "testsuntonemap.png",
			--sunrise = "default_stone.png",
			sunrise_visible = false,
			scale = 4
		})--]]

		--[[
		player:set_moon({
			visible = true,
			--texture = "default_stone.png",
			--tonemap = "testmoontonemap.png",
			scale = 4
		})--]]

		--[[
		player:set_stars({
			visible = false,
			count = 4000,
			star_color = "#ff00ff69",
			size = 2
		})--]]

		--[[ old call
		player:set_sky(
			--"#000000",
			{r = 0, g = 0, b = 0, a = 0},
			"skybox",
			--"plain",
			{
				"default_sand.png",
				"default_stone.png",
				"default_grass.png",
				"default_jungletree.png",
				"default_obsidian.png",
				"default_lava.png"
			},
			true
		)--]]
	end
	minetest.after(5, cyclic_update)
end

minetest.after(5, cyclic_update)
```

sfan5 commented:
Full merging instructions (equal to a rebase & squash, there will be no merge commit):
<pre>
git fetch origin pull/8964/merge
git merge FETCH_HEAD -s recursive -X theirs
git reset origin/master
git add -u && git add src/skyparams.h
git commit --author="Jordach &lt;jordach.snelling@gmail.com&gt;" --date="Wed, 21 Aug 2019 21:47:45 +0100" -m "set_sky improvements, set_sun, set_moon and set_stars"
</pre>

Jordach commented:
> Custom fog moon / sun tint creates pulatsing fog tint at dusk/dawn.

I'd have to take a look at that seems like it's possibly *not* working with a tone map enabled.

paramat commented:
If it helps, i did not have a custom tonemap specified at the time.

On that subject, i did not test the 'tonemap' feature in 'set sun' or 'set moon'. Cannot be bothered due to the effort needed, and because i dislike the existing tonemap feature and consider it overkill and unnecessary (but i do not oppose it in your PR).

Looking forward to this being merged.

Jordach commented:
Tonemaps are a bit unique, in that they can actually colorise the sun and moon without using a lot of memory due to the amount of textures required by abusing `set_sun({texture=...})` while a tonemap can do this by changing the color of the actual mesh sun. Reducing memory footprint.

SmallJoker commented on src/network/clientpackethandler.cpp:
```diff
@@ -1240,28 +1241,69 @@ void Client::handleCommand_HudSetParam(NetworkPacket* pkt)
+	u16 texture_count;
+	std::string texture;
+
+	*pkt >> skybox.bgcolor >> skybox.type >> skybox.clouds >>
```
Beginning from `skybox.clouds` it's not backwards compatible. The absolute minimal compatibility code would look as follows:
```C++
if (m_proto_ver < 39) {
	// Skybox not applied on older servers
	return;
}
// New skybox format
*pkt >> skybox.bgcolor >> skybox.type >> ....
```

Jordach replied:
Would leaving a warning in console be wise enough or actually going the whole way and keeping compatibility?

sfan5 replied:
Full compatibility is definitely needed here, you won't want the skybox to go missing just because you connected to an older server.

Jordach replied:
Coming back to this, should an old packet keep the classic intended set_sky behaviour, as in hiding the sun, moon and stars depending on skybox mode?

sfan5 replied:
It should come reasonably close to the behavior the server actually intended.
That's a yes.

Jordach replied:
It's fully functional with 5.1.0-dev from local testing, works on Inside The Box.

SmallJoker commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1701,43 +1701,183 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
+			is_colorspec = lua_isnumber(L, -1);
+		lua_getfield(L, 2, "b");
+		if (!is_colorspec)
+			is_colorspec = lua_isnumber(L, -1);
```
Please use `read_ARGB8` here or move this `ColorSpec`-specific code to `c_converter.cpp`.

Jordach replied:
Probably worth moving to c_converter.cpp as this is a read only, non stack-popping operation.

SmallJoker commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1701,43 +1701,183 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
 		}
-	}
+		getServer(L)->setSky(player, skybox_params);
+		lua_pushboolean(L, true);
```
`return 1;` here and dropping the `else` below would make the code and diff easier to read.

Jordach replied:
Are you sure? It'd be out of scope as SkyboxParams isn't used outside of it's functions.

Jordach replied:
Nevermind, I got what you mean, move SkyboxParams outside of the if else statement block.

SmallJoker commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1701,43 +1701,183 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
+		lua_getfield(L, 2, "textures");
+		skybox_params.textures.clear();
+		if (!lua_istable(L, -1)) {
+		} else {
```
`if (lua_istable(L, -1) && skybox_params.type == "skybox") {`
This would replace the empty case above and ignore the unused `textures` parameter for non-skybox modes.

SmallJoker commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1701,43 +1701,183 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
+		lua_getfield(L, 2, "sky_color");	
+		if (lua_istable(L, -1)) {
+			lua_getfield(L, -1, "day_sky");
+			if (!lua_isnil(L, -1))
```
Superflous `lua_isnil` checks. `read_color` already handles this case (color is not updated).

SmallJoker commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1747,26 +1887,235 @@ int ObjectRef::l_get_sky(lua_State *L)
+	lua_setfield(L, -2, "tint_type");
+	lua_setfield(L, -2, "sky_color");
+	
+	return 1;
```
This breaks the API. Mods expect this function to return 4 values.
Suggestion: keep `[gs]et_sky()` for basic changes, and introduce `[gs]et_skybox` for full skybox controls.

Jordach replied:
I could remove the named groups for the old stuff, but move new stuff into it's own table, or just make `set_skybox()` and `get_skybox()`

SmallJoker replied:
Yes, that would be a mod-compatible solution.

Jordach replied:
What I've done instead of trying the first, I've gone with adding `player:get_sky_color()` which returns the table as found here. `player:get_sky()` now returns the four values: fog/bgcolor, type, textures, clouds.

SmallJoker commented:
I did some testing with `atmos`, but I cannot figure out why the clouds flicker shortly every 5...30 seconds.
Here's a basic testing mod for different `set_sky` configurations:

<details>
<summary>Some testing code</summary>

```Lua
local defs = {
	plain = {
		base_color = "red",
		type = "plain",
		textures = { "ignored", "params", "ok", "foo", "bar", "test" },
		clouds = true
	},
	regular = {
		type = "regular",
		clouds = true,
		sky_color = {
			day_sky = "#AAF",
			day_horizon = "#55F",
			dawn_sky = "#FAA",
			dawn_horizon = "#F55",
			night_sky = "#AFA",
			night_horizon = "#5F5",
			indoors = "#000",
			fog_sun_tint = "#FF6",
			fog_moon_tint = "#F6A",
			--fog_tint_type = "default",
			fog_tint_type = "custom",
		}
	},
	noclouds_regular = {
		type = "regular",
		clouds = false
	},
	noclouds_plain = {
		base_color = "red",
		type = "plain",
		clouds = false
	},
	skybox = {
		type = "skybox",
		textures = {
			"default_dirt.png", "default_sand.png",
			"default_wood.png", "default_stone.png",
			"default_cobble.png", "default_brick.png"
		},
		clouds = true,
	},
	skybox_err = {
		type = "skybox",
		clouds = false,
		textures = {}
	}
}

minetest.register_chatcommand("sky", {
	func = function(name, param)
		local player = minetest.get_player_by_name(name)
		if not defs[param] then
			param = "plain"
		end
		player:set_sky(defs[param])
		return true, "Set sky to " .. param
	end
})
```
</details>

`skybox_err` is especially interesting because it's showing a "random" color after each world join 8not a serious bug).

sfan5 commented on src/client/sky.cpp:
```diff
@@ -737,16 +816,199 @@ void Sky::draw_sky_body(std::array<video::S3DVertex, 4> &vertices, float pos_1,
+	if (m_sun_tonemap)
+		m_materials[3].Lighting = true;
+	else
+		m_materials[3].Lighting = false;
```
Instead of casting, `m_materials[3].Lighting = !!m_sun_tonemap;` is also an option

Jordach commented on src/client/sky.cpp:
```diff
@@ -737,16 +816,199 @@ void Sky::draw_sky_body(std::array<video::S3DVertex, 4> &vertices, float pos_1,
+	* vertices: The body as a rectangle of 4 vertices
+	* horizon_position: turn the body around the Y axis
+	* day_position: turn the body around the Z axis, to place it depending of the time of the day
+	*/
```
Oh it appears MSVC clang-format removed the spaces used to make the asterisks line up at the start of the comment block.

Jordach commented on src/client/sky.cpp:
```diff
@@ -737,16 +816,199 @@ void Sky::draw_sky_body(std::array<video::S3DVertex, 4> &vertices, float pos_1,
+	// Keep a list of texture names handy.
+	m_sky_params.textures.emplace_back(texture);
+	video::ITexture *result = tsrc->getTextureForMesh(texture);
+	m_materials[material_id+5] = m_materials[0];
```
I suppose `i > 5` would fit the bill as it's indexed from zero in game.cpp.

SmallJoker replied:
Sorry? It does not matter (much) where this function is called. It should be safe to not corrupt memory.

Jordach replied:
I've added a small check to check for invalid material ids that exist past 11.

Jordach commented on src/client/sky.h:
```diff
@@ -79,12 +94,19 @@ class Sky : public scene::ISceneNode
+	void setSkyColors(video::SColor bg_day, video::SColor bg_dawn,
+		video::SColor bg_night, video::SColor sky_day,
+		video::SColor sky_dawn, video::SColor sky_night,
+		video::SColor bg_indoor);
```
Probably worth the effort, seeing as I managed it with sky.h

Jordach replied:
I can't think of a straightforwards way of doing it not without making it look like a mess.

SmallJoker replied:
Currently it's a mess, and I'm pretty sure reducing the input arguments will improve it.

Jordach replied:
Ah, I solved it by moving the sky colors to a struct in a struct allowing for the sky colors to be fully replaced. The best part is I can do `m_sky_params.sky_color = sky.sky_color`

Jordach commented:
Testing of protocol 38 server backwards compatibility with old set_sky packets:
https://gfycat.com/ornerybrightjavalina

Jordach commented:
Flickering clouds is caused by constantly resetting the values used for calculating the sun/moon tint at their resepective rise and set times.

Jordach commented on src/client/sky.cpp:
```diff
@@ -558,13 +520,17 @@ void Sky::update(float time_of_day, float time_brightness,
 				pointcolor_sun_f.b = pointcolor_light *
 					(float)m_materials[3].EmissiveColor.getBlue() / 255;
 				pointcolor_sun_f.g = pointcolor_light *
 					(float)m_materials[3].EmissiveColor.getGreen() / 255;
```
I have a suspect feeling this is the bug in question.

Jordach replied:
Accidential, found it.

Jordach commented on src/client/sky.cpp:
```diff
@@ -573,9 +539,23 @@ void Sky::update(float time_of_day, float time_brightness,
+					m_sky_params.moon_tint.getRed() * pointcolor_light,
+					m_sky_params.moon_tint.getGreen() * pointcolor_light,
+					m_sky_params.moon_tint.getBlue() * pointcolor_light,
+					1
```
The offending bug;

get[Red/Green/Blue] returns 0-255; not 0 - 1.0f.

The fix is to divide the results by 255, then multiply by `pointcolor_light`.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -5832,15 +5832,82 @@ This is basically a reference to a C++ `ServerActiveObject`
+            * `"skybox"`: Uses 6 textures, `base_color` used as fog.
+            * `"plain"`: Uses 0 textures, `base_color` used as both fog and sky.
+        * `textures`: A table containing up to six textures in the following
+            order: top, bottom, left, right, front, back.
```
Where's `left`, `right`, ..? I assume that's with rotation = 0°, yet only an assumption.
Please use `X+`, `Z-`, ..  to describe the directions.

Jordach replied:
Described as Y+ (top), Y- (bottom), X- (west), X+ (east), Z+ (north), Z- (south)

SmallJoker commented on src/client/game.cpp:
```diff
@@ -2744,41 +2750,85 @@ void Game::handleClientEvent_HudChange(ClientEvent *event, CameraOrientation *ca
+			event->set_sky->tint_type
+		);
+	} else if (event->set_sky->type == "skybox" &&
+		event->set_sky->textures.size() == 6) {
```
+1 tabulator

SmallJoker commented on src/client/sky.cpp:
```diff
@@ -67,44 +68,52 @@ Sky::Sky(s32 id, ITextureSource *tsrc):
 	}
-
 	m_directional_colored_fog = g_settings->getBool("directional_colored_fog");
+	m_star_params.count = 1;
```
Hacky implementation. Please add a boolean parameter to `setStarCount` to force updating the stars if that's really needed.

SmallJoker commented on src/client/sky.cpp:
```diff
@@ -738,15 +818,164 @@ void Sky::draw_sky_body(std::array<video::S3DVertex, 4> &vertices, float pos_1,
+		ITextureSource *tsrc)
+{
+	// Sanity check for more than six textures.
+	if (material_id > 5)
```
`if (material_id + 5 >= SKY_MATERIAL_COUNT)`

SmallJoker commented on src/network/clientpackethandler.cpp:
```diff
@@ -1240,28 +1241,130 @@ void Client::handleCommand_HudSetParam(NetworkPacket* pkt)
-	std::vector<std::string> *params = new std::vector<std::string>;
+		SkyboxParams skybox;
+		skybox.bgcolor = video::SColor(readARGB8(is));
+		skybox.type = std::string(deSerializeString(is));
```
What about the other struct variables? They're not initialized properly.

sfan5 replied:
Doesn't C++ call the default constructor when you declare a struct?

Jordach replied:
Setting `skybox.tint_type = "default"` doesn't require `sun_tint` or `moon_tint` to be initialised, as sky.cpp handles this special case, (or to use tonemapping when enabled), to use the original sun and moon tinting at dawn/dusk.

SmallJoker replied:
I should look better at the actual struct variables. Only `clouds` needs to be initialized manually.

Jordach replied:
I've set `sun_tint` and `moon_tint` to `RGBA(255, 255, 255, 255)` just in case *something* goes wrong.

SmallJoker commented on src/network/clientpackethandler.cpp:
```diff
@@ -1240,28 +1241,130 @@ void Client::handleCommand_HudSetParam(NetworkPacket* pkt)
+		try {
+			skybox.clouds = readU8(is);
+		}
+		catch (...) {}
```
Please put `catch` and `else` (below) after `}`.

SmallJoker commented on src/client/sky.cpp:
```diff
@@ -717,14 +703,108 @@ void Sky::draw_moon(video::IVideoDriver *driver, float moonsize, const video::SC
+	float f = starbrightness;
+	float d = (0.006 / 2) * m_star_params.scale;
+
+	video::SColor starcolor(f * m_star_params.starcolor.getAlpha(),
```
Shorter:
```C++
video::SColor starcolor = m_star_params.starcolor;
starcolor.setAlpha(f * m_star_params.starcolor.getAlpha());

// Stars are only drawn when not fully transparent
if (starcolor < 1)
	return;
```

Jordach replied:
I have the succinct feeling that `if (starcolor < 1)` would be invalid as the alpha channel is used to determine whether the stars aren't *fully invisible* due to alpha blending.

SmallJoker commented on src/network/clientpackethandler.cpp:
```diff
@@ -1240,28 +1241,130 @@ void Client::handleCommand_HudSetParam(NetworkPacket* pkt)
+		moon_event->moon_params = new MoonParams(moon);
+		m_client_event_queue.push(moon_event);
+
+		ClientEvent* star_event = new ClientEvent();
```
Please move the pointer towards `(.*)_event`.

SmallJoker commented on src/network/networkprotocol.h:
```diff
@@ -602,14 +605,22 @@ enum ToClientCommand
-		foreach count:
-			u8 len
-			u8[len] param
-		u8 clouds (boolean)
```
The old format documentation should be kept until the compatibility code is dropped.

SmallJoker commented on src/script/common/c_converter.cpp:
```diff
@@ -335,6 +335,28 @@ video::SColor read_ARGB8(lua_State *L, int index)
+	// Check whole table in case of missing ColorSpec keys:
+	// This check does not remove the checked value from the stack.
+	// Only update the value if we know we have a valid ColorSpec key pair.
+	if (lua_istable(L, index)) {
```
```C++
if (!lua_istable(L, index))
	return false;
```

Reduces the amount of indents, and gets rid of the useless `else` case.

SmallJoker commented on src/script/common/c_converter.cpp:
```diff
@@ -335,6 +335,28 @@ video::SColor read_ARGB8(lua_State *L, int index)
 	return color;
 }
 
+bool is_color_table(lua_State* L, int index)
```
Pointer needs correction, in header too.

SmallJoker commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1747,28 +1862,226 @@ int ObjectRef::l_get_sky(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef* ref = checkobject(L, 1);
+	RemotePlayer* player = getplayer(ref);
```
Pointer placement. L1885 too.

SmallJoker commented on src/skyparams.h:
```diff
@@ -0,0 +1,114 @@
+class SkyboxDefaults
+{
+public:
+	const skycolor getSkyColorDefaults() {
```
Please move the bracket to the next line. Same for the other functions below.

sfan5 commented on src/script/common/c_converter.cpp:
```diff
@@ -335,6 +335,28 @@ video::SColor read_ARGB8(lua_State *L, int index)
+		if (!is_color_table)
+			is_color_table = lua_isnumber(L, -1);
+		return is_color_table;
+	}
```
the else clause is redundant since it returns early

sfan5 commented:
@SmallJoker this PR appears to be ready for re-review

SmallJoker commented on src/client/sky.cpp:
```diff
@@ -717,14 +702,106 @@ void Sky::draw_moon(video::IVideoDriver *driver, float moonsize, const video::SC
+	// to time 4000.
+
+	float tod = wicked_time_of_day < 0.5f ? wicked_time_of_day : (1.0f - wicked_time_of_day);
+	float starbrightness = clamp((0.25f - fabs(tod)) * 20.0f, 0.0f, 1.0f);
```
```
[...]/src/client/sky.cpp:715:25: error: no matching function for call to 'clamp' [clang-diagnostic-error]

        float starbrightness = clamp((0.25f - fabs(tod)) * 20.0f, 0.0f, 1.0f);
```
Worth fixing clang. https://travis-ci.org/minetest/minetest/jobs/654161899

sfan5 replied:
`fabs` -> `fabsf` should do the trick

I can also do this on merge

sfan5 commented:
merged as 946c03c69bfdde7dc91295692479f8e81bdf79e9

appgurueu commented:
Can't celebrate enough that this has been merged :confetti_ball:

paramat commented:
Thanks for everyone's hard work on this, so good to see it merged, an excellent feature.
